File: typescript/src/schemas.ts
import { z } from 'zod';

// Mirrors CreateUrlRequest in blueprint/contracts/openapi.yaml, including
// additionalProperties: false.
export const CreateUrlRequestSchema = z
  .object({
    original_url: z.string(),
    expires_at: z.string().optional(),
  })
  .strict();

// Mirrors the ShortUrl schema in blueprint/contracts/openapi.yaml.
export const ShortUrlSchema = z.object({
  code: z.string().regex(/^[A-Za-z0-9]{7}$/),
  original_url: z.string(),
  short_url: z.string(),
  click_count: z.number().int().min(0),
  created_at: z.string(),
  expires_at: z.string().nullable(),
});

export const ValidationErrorSchema = z.object({ detail: z.string() });
export const NotFoundErrorSchema = z.object({ detail: z.literal('Short URL not found') });
export const ExpiredErrorSchema = z.object({ detail: z.literal('Short URL has expired') });

export type ShortUrl = z.infer<typeof ShortUrlSchema>;

File: rust/src/schemas.rs
//! Request/response DTOs mirroring the schemas in
//! `blueprint/contracts/openapi.yaml`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::timestamps;

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CreateUrlRequest {
    pub original_url: Option<String>,
    pub expires_at: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ShortUrl {
    pub code: String,
    pub original_url: String,
    pub short_url: String,
    pub click_count: u64,
    #[serde(serialize_with = "timestamps::serialize")]
    pub created_at: DateTime<Utc>,
    #[serde(serialize_with = "timestamps::serialize_opt")]
    pub expires_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize)]
pub struct ErrorBody {
    pub detail: String,
}

impl ErrorBody {
    pub fn new(detail: impl Into<String>) -> Self {
        Self { detail: detail.into() }
    }
}
